File: app/src/main/java/com/bitflaker/lucidsourcekit/main/goals/RecyclerViewAdapterCurrentGoals.java
package com.bitflaker.lucidsourcekit.main.goals;

import android.content.Context;
import android.content.res.ColorStateList;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bitflaker.lucidsourcekit.R;
import com.bitflaker.lucidsourcekit.database.MainDatabase;
import com.bitflaker.lucidsourcekit.database.goals.entities.Goal;
import com.bitflaker.lucidsourcekit.database.goals.entities.ShuffleTransaction;
import com.bitflaker.lucidsourcekit.databinding.EntryCurrentGoalBinding;
import com.bitflaker.lucidsourcekit.databinding.SheetGoalTransactionsBinding;
import com.bitflaker.lucidsourcekit.utils.Tools;
import com.google.android.material.bottomsheet.BottomSheetDialog;

import java.util.Calendar;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

public class RecyclerViewAdapterCurrentGoals extends RecyclerView.Adapter<RecyclerViewAdapterCurrentGoals.MainViewHolder> {
    private final Context context;
    private final List<Goal> goals;
    private final int shuffleId;
    private final MainDatabase db;
    private final HashMap<Integer, Integer> goalAchievedCount;
    private final ColorStateList colorNotAchieved, colorAchieved;

    public RecyclerViewAdapterCurrentGoals(Context context, List<Goal> goals, int shuffleId) {
        this.context = context;
        this.goals = goals;
        this.shuffleId = shuffleId;
        db = MainDatabase.getInstance(context);
        goalAchievedCount = db.getShuffleTransactionDao().getAllCountsFromShuffle(shuffleId)
                .blockingGet()
                .stream()
                .collect(Collectors.toMap(x -> x.goalId, x -> x.count, (prev, next) -> next, HashMap::new));
        colorNotAchieved = Tools.getAttrColorStateList(R.attr.colorSurfaceContainer, context.getTheme());
        colorAchieved = Tools.getAttrColorStateList(R.attr.colorTertiary, context.getTheme());
    }

    @NonNull
    @Override
    public RecyclerViewAdapterCurrentGoals.MainViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(context);
        return new RecyclerViewAdapterCurrentGoals.MainViewHolder(EntryCurrentGoalBinding.inflate(inflater, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerViewAdapterCurrentGoals.MainViewHolder holder, int position) {
        Goal currentGoal = goals.get(position);
        holder.binding.txtGoal.setText(currentGoal.description);

        int achievedCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0);
        holder.binding.imgAchievedCounterBackground.setImageTintList(achievedCount == 0 ? colorNotAchieved : colorAchieved);
        holder.binding.txtAchievedCounter.setVisibility(achievedCount == 0 ? View.GONE : View.VISIBLE);
        holder.binding.txtAchievedCounterPlaceholder.setVisibility(achievedCount == 0 ? View.VISIBLE : View.GONE);
        holder.binding.txtAchievedCounter.setText(String.format(Locale.getDefault(), "%d", achievedCount));
        holder.binding.crdCurrentGoal.setOnClickListener(e -> {
            ShuffleTransaction transaction = new ShuffleTransaction(shuffleId, currentGoal.goalId, Calendar.getInstance().getTimeInMillis());
            db.getShuffleTransactionDao().insert(transaction).blockingSubscribe();
            int newCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0) + 1;
            goalAchievedCount.put(currentGoal.goalId, newCount);
            holder.binding.txtAchievedCounter.setText(String.format(Locale.getDefault(), "%d", newCount));
            if (achievedCount == 0) {
                notifyItemChanged(position);
            }
        });
        holder.binding.crdCurrentGoal.setOnLongClickListener(e -> {
            final BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(context, R.style.BottomSheetDialogStyle);
            SheetGoalTransactionsBinding sBinding = SheetGoalTransactionsBinding.inflate(LayoutInflater.from(context));
            bottomSheetDialog.setContentView(sBinding.getRoot());

            List<ShuffleTransaction> transactions = db.getShuffleTransactionDao().getAllFromShuffleGoal(shuffleId, currentGoal.goalId).blockingGet();
            RecyclerViewAdapterGoalTransactions manager = new RecyclerViewAdapterGoalTransactions(context, transactions);
            sBinding.rcvShuffleTransactions.setAdapter(manager);
            sBinding.rcvShuffleTransactions.setLayoutManager(new LinearLayoutManager(context));
            sBinding.txtGoal.setText(currentGoal.description);
            sBinding.txtNoneAchieved.setVisibility(transactions.isEmpty() ? View.VISIBLE : View.GONE);
            sBinding.rcvShuffleTransactions.setVisibility(transactions.isEmpty() ? View.GONE : View.VISIBLE);

            manager.setOnEntryDeletedListener(deleted -> {
                int newCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0) - 1;
                goalAchievedCount.put(currentGoal.goalId, newCount);
                notifyItemChanged(position);
                if (transactions.isEmpty()) {
                    sBinding.txtNoneAchieved.setVisibility(View.VISIBLE);
                }
            });

            bottomSheetDialog.show();
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return goals.size();
    }

    public static class MainViewHolder extends RecyclerView.ViewHolder {
        EntryCurrentGoalBinding binding;

        public MainViewHolder(@NonNull EntryCurrentGoalBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/bitflaker/lucidsourcekit/main/goals/views/RecyclerViewAdapterCurrentGoals.kt
package com.bitflaker.lucidsourcekit.main.goals.views

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bitflaker.lucidsourcekit.R
import com.bitflaker.lucidsourcekit.database.MainDatabase
import com.bitflaker.lucidsourcekit.database.goals.entities.Goal
import com.bitflaker.lucidsourcekit.database.goals.entities.ShuffleTransaction
import com.bitflaker.lucidsourcekit.databinding.EntryCurrentGoalBinding
import com.bitflaker.lucidsourcekit.databinding.SheetGoalTransactionsBinding
import com.bitflaker.lucidsourcekit.utils.attrColorStateList
import com.google.android.material.bottomsheet.BottomSheetDialog
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Locale

class RecyclerViewAdapterCurrentGoals(
    private val fragment: Fragment,
    private val goals: List<Goal>?,
    private val shuffleId: Int
) : RecyclerView.Adapter<RecyclerViewAdapterCurrentGoals.MainViewHolder>() {

    class MainViewHolder(var binding: EntryCurrentGoalBinding) : RecyclerView.ViewHolder(binding.root)

    private lateinit var goalAchievedCount: HashMap<Int, Int>
    private val activity = fragment.requireActivity()
    private val db: MainDatabase = MainDatabase.getInstance(activity)
    private val colorNotAchieved = activity.attrColorStateList(R.attr.colorSurfaceContainer)
    private val colorAchieved = activity.attrColorStateList(R.attr.colorTertiary)

    init {
        // Load goal achieved counts and then update entries
        fragment.lifecycleScope.launch(Dispatchers.IO) {
            goalAchievedCount = db.shuffleTransactionDao.getAllCountsFromShuffle(shuffleId)
                .groupBy { it.goalId }
                .mapValues { it.value.single().count }
                .toMap(HashMap())

            activity.runOnUiThread {
                notifyItemRangeChanged(0, itemCount)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MainViewHolder {
        val binding = EntryCurrentGoalBinding.inflate(LayoutInflater.from(activity), parent, false)
        return MainViewHolder(binding)
    }

    override fun onBindViewHolder(holder: MainViewHolder, position: Int) {
        val currentGoal = goals?.get(position) ?: return
        holder.binding.txtGoal.text = currentGoal.description

        val achievedCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0)
        holder.binding.imgAchievedCounterBackground.setImageTintList(if (achievedCount == 0) colorNotAchieved else colorAchieved)
        holder.binding.txtAchievedCounter.visibility = if (achievedCount == 0) View.GONE else View.VISIBLE
        holder.binding.txtAchievedCounterPlaceholder.visibility = if (achievedCount == 0) View.VISIBLE else View.GONE
        holder.binding.txtAchievedCounter.text = String.format(Locale.getDefault(), "%d", achievedCount)

        // Set click listener to increase the achieved count of the current goal by 1
        holder.binding.crdCurrentGoal.setOnClickListener {
            fragment.lifecycleScope.launch(Dispatchers.IO) {
                val transaction = ShuffleTransaction(shuffleId, currentGoal.goalId, Calendar.getInstance().timeInMillis)
                db.getShuffleTransactionDao().insert(transaction)
                val newCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0) + 1
                goalAchievedCount.put(currentGoal.goalId, newCount)

                activity.runOnUiThread {
                    // TODO: This could theoretically cause issues, but probably won't
                    //       in practice (Changing item that might have moved already)
                    //       This is to make it more responsive and prevent the change
                    //       animation in this scenario
                    holder.binding.txtAchievedCounter.text = String.format(Locale.getDefault(), "%d", newCount)
                    if (achievedCount == 0) {
                        notifyItemChanged(position)
                    }
                }
            }
        }

        // Set long click listener for showing editor of achieved timestamps and counts
        holder.binding.crdCurrentGoal.setOnLongClickListener {
            val bottomSheetDialog = BottomSheetDialog(activity, R.style.BottomSheetDialogStyle)
            val sBinding = SheetGoalTransactionsBinding.inflate(LayoutInflater.from(activity))
            bottomSheetDialog.setContentView(sBinding.root)

            // Load achieved goal times
            fragment.lifecycleScope.launch(Dispatchers.IO) {
                val transactions = db.getShuffleTransactionDao().getAllFromShuffleGoal(shuffleId, currentGoal.goalId)

                activity.runOnUiThread {
                    val manager = RecyclerViewAdapterGoalTransactions(fragment, transactions)
                    sBinding.rcvShuffleTransactions.setAdapter(manager)
                    sBinding.rcvShuffleTransactions.setLayoutManager(LinearLayoutManager(activity))
                    sBinding.txtGoal.text = currentGoal.description
                    sBinding.txtNoneAchieved.visibility = if (transactions.isEmpty()) View.VISIBLE else View.GONE
                    sBinding.rcvShuffleTransactions.visibility = if (transactions.isEmpty()) View.GONE else View.VISIBLE

                    // Set listener to refresh entry when an entry was removed
                    manager.onDeletedListener = {
                        val newCount = goalAchievedCount.getOrDefault(currentGoal.goalId, 0) - 1
                        goalAchievedCount.put(currentGoal.goalId, newCount)
                        notifyItemChanged(position)
                        if (transactions.isEmpty()) {
                            sBinding.txtNoneAchieved.visibility = View.VISIBLE
                        }
                    }

                    bottomSheetDialog.show()
                }
            }

            true
        }
    }

    override fun getItemCount(): Int {
        return goals?.size ?: 0
    }
}
